<a href="https://colab.research.google.com/github/DeepuBenson/Python/blob/main/Day1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Python: A multi-paradigm programming language**

A few examples for functions in Python

In [2]:
def intsum(n):
    return sum(range(n+1))

print(intsum(100))

5050


In [3]:
def foo(n):
    print(n+n)

In [4]:
foo('1234')
foo(1234)
foo([1,2,3,4])
foo((1,2,3,4))

12341234
2468
[1, 2, 3, 4, 1, 2, 3, 4]
(1, 2, 3, 4, 1, 2, 3, 4)


In [5]:
n=int(input('Enter a positive number: '))
if n > 1:
    for i in range(2,n-1):
        if n % i == 0:
            print(n, "is composite")
            break
    else:     #Python for loops can also have else!                      
        print(n, "is prime")
else:         
    print(n, "is neither prime nor composite")

Enter a positive number: 123
123 is composite


In [6]:
def intsum(n):
    if n==1:
        return 1
    else:
        return n+intsum(n-1)

x=int(input('Enter a number : '))
print(intsum(x))

Enter a number : 122
7503


In [7]:
def fact(n):
    if n==1:
        return 1
    else:
        return n*fact(n-1)

x=int(input('Enter a number : '))
print(fact(x))

Enter a number : 6
720


A few examples for classes in Python

In [8]:
class ClassName:
    #Body of the class
    #should not be empty
    pass 

In [11]:
class OurClass:
    x=123

oc1=OurClass() #First Object created
oc2=OurClass() #Second Object created

print(oc1.x)
print(oc2.x)

oc1.x=456

print(oc1.x)   # ???
print(oc2.x)   # ???

print(id(oc1)) # ???
print(id(oc2)) # ???

123
123
456
123
140611706230624
140611706230432


In [12]:
class Check:
    a=100
    def __init__(self,n): #Similar to a constructor in C++/Java
        self.n = n
    def checking(self):   # This is called a method
        print("---->",Check.a)
        if self.n % 2 == 0:
            print(self.n,'is an even number')
        else:
            print(self.n,'is an odd number')

    def ch(self):
        Check.a=333

x=Check(123)
y=Check(456)
x.ch()
x.checking()
y.checking()

----> 333
123 is an odd number
----> 333
456 is an even number


Functional Programming with Python

In [13]:
lam = lambda n: n+n

In [14]:
lam(123)

246

In [15]:
lam('123')

'123123'

In [16]:
def foo(n):
    return lambda x: x**n

In [17]:
square = foo(2)

In [18]:
print(square)

<function foo.<locals>.<lambda> at 0x7fe2b9f23b80>


In [19]:
square(5)

25

In [20]:
cube = foo(3)

In [21]:
cube(5)

125

In [25]:
x='123'
print(int(x))

123


In [40]:
y=['11', '22', '33', '44', '55']

In [41]:
print(int(y))   #Intentional Mistake!

TypeError: ignored

In [42]:
print(list(map(int,y)))

[11, 22, 33, 44, 55]


In [45]:
import math as m
y = [(x*m.pi)/180 for x in range(0, 361,30)] # Example for List Comprehension. A very useful feature.
print(y)

[0.0, 0.5235987755982988, 1.0471975511965976, 1.5707963267948966, 2.0943951023931953, 2.6179938779914944, 3.141592653589793, 3.6651914291880923, 4.1887902047863905, 4.71238898038469, 5.235987755982989, 5.759586531581287, 6.283185307179586]


In [50]:
print(list(map(m.sin, y)))

[0.0, 0.49999999999999994, 0.8660254037844386, 1.0, 0.8660254037844387, 0.49999999999999994, 1.2246467991473532e-16, -0.5000000000000001, -0.8660254037844384, -1.0, -0.8660254037844386, -0.5000000000000004, -2.4492935982947064e-16]


**Why Sin(30) is shown as 0.49999999999999994 instead of 0.5?**

A long, detailed and difficult explanation can be found here ---> [What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)

**Optional reading material**

In [46]:
print(m.sqrt(3)/2)

0.8660254037844386


In [48]:
def foo(n):
    return n+n

In [49]:
x=[1, 2, 3, 4, 5]

In [51]:
print(foo(x))

[1, 2, 3, 4, 5, 1, 2, 3, 4, 5]


In [53]:
print(list(map(foo, x)))

[2, 4, 6, 8, 10]


In [54]:
print(list(map(lambda n: n+n, x)))

[2, 4, 6, 8, 10]


In [60]:
a = [1, 2, 3, 4]
b = [4, 5, 6, 7]
print(list(map(lambda x, y: x * y, a, b)))

[4, 10, 18, 28]


In [61]:
print(set(map(lambda x, y: x * y, a, b)))

{10, 18, 4, 28}


In [71]:
from random import randint
rlist = [randint(1,100) for x in range(10)]
print(rlist)

[69, 70, 74, 88, 38, 100, 16, 34, 24, 84]


In [72]:
print(list(filter(lambda x: x%2 == 1, rlist)))

[69]


In [76]:
x=['cat', 'boat', 'mat', 'sat', 'bat', 'ball', 'rat']
print(list(filter(lambda a: a.startswith('b'), x)))

['boat', 'bat', 'ball']


In [77]:
y = [11, {}, (), 22, '', 33, None, 0, 0.0, 44, "", 55]
print(y)

[11, {}, (), 22, '', 33, None, 0, 0.0, 44, '', 55]


In [79]:
print(list(filter(None, y)))

[11, 22, 33, 44, 55]


In [80]:
print(list(filter(lambda a: a, y)))

[11, 22, 33, 44, 55]


In [82]:
from functools import reduce
lst = [11, 44, 66, 88, 33, 22, 77]
print(reduce(lambda a, b: a if a >=b else b, lst))

88


In [83]:
max(lst)

88

In [86]:
lst = [1, 2, 3, 4, 5, 6]

In [87]:
print(reduce(lambda a, b: a*b, lst))

720


In [88]:
import operator as op

print(reduce(op.mul, lst))

720


**Practice Problem 1**: Find the sum of the squares of the even numbers in a list.

In [90]:
# I will call this a C program written in Python. Why?

x=[11,12,13,14,15,16,17,18,19,20]
s=0
for i in x:
	if i % 2 ==0:
		s=s+i**2 
print(s)


1320


In [91]:
# Do you have a "Pythonic" solution. Try here!




**Practice Problem 2**: Read a number from the keyboard and print the sum of its digits.

In [95]:
# Another bad Python script.

n=int(input('Enter a number: '))
s=0
while(n > 0):
	s=s+n%10
	n=n//10
print(s)

Enter a number: 123456
21


In [94]:
# Do you have a "Pythonic" solution. Try here!




**Practice Problem 3:** From a sentence, find words of length greater than n. 

In [97]:
# One more bad Python script

str=input("Enter a sentence: ")
st=str.split()
n=int(input("Enter the length: "))
nstr=[]
for i in st:
	if len(i) > n:
		nstr=nstr+[i]
print("Words with length greater than", n, "are",nstr)

Enter a sentence: quick fox had a rabbit
Enter the length: 8
Words with length greater than 8 are []


In [98]:
# Do you have a "Pythonic" solution. Try here!


